src/metabolic_ninja/helpers.py

```diff
@@ -0,0 +1,178 @@
+        solver tolerance.
+
+    Returns
+    -------
+    list
+        All the candidate exotic (non-native) co-factors.
+
+    Raises
+    ------
+    OptimizationError
+        If producing the heterologous product with the given model is
+        infeasible.
+
+    Warnings
+    --------
+    The current design is only meant for linear pathways. It will likely be
+    incorrect for branching ones.
+
+    """  # noqa: E501
+    # Find the direction of flux through the pathway to the product.
+    with model:
+        pathway.apply(model)
+        model.objective = model.reactions.get_by_id(pathway.product.id)
+        model.objective_direction = "max"
+        solution = model.optimize()
+    # Adapters are single compound reactions: native <=> heterologous.
+    adapted_sources = frozenset([r.products[0] for r in pathway.adapters])
+    heterologous_reactions = frozenset(pathway.reactions)
+    # The pathway `product` is a demand reaction.
+    target = pathway.product.reactants[0]
+    atom_counts = {
+        target: count_atoms(target.formula)
+    }
+    # We are interested in co-factors of real heterologous reactions only, i.e.,
+    # not adapters nor exchanges.
+    rxn_queue = list(target.reactions & heterologous_reactions)
+    seen = set()
+    exotic_cofactors = set()
+    while len(rxn_queue) > 0:
+        rxn = rxn_queue.pop()
+        seen.add(rxn)
+        flux = 0.0 if abs(solution[rxn.id]) <= threshold else solution[rxn.id]
+        if flux > 0:
+            substrates = set(rxn.reactants)
+            products = set(rxn.products)
+        elif flux < 0:
+            substrates = set(rxn.products)
+            products = set(rxn.reactants)
+        else:
+            logger.warning("No flux through heterologous reaction %r!",
+                           rxn.id)
+            continue
+        products.remove(target)
+        exotic_cofactors.update(products - adapted_sources)
+        # With one precursor, there is no doubt that it is not a co-factor.
+        if len(substrates) == 1:
+            target = substrates.pop()
+            rxn_queue.extend((target.reactions & heterologous_reactions) - seen)
+            continue
+        # With multiple precursors, we evaluate the strength of chemical
+        # linkage in order to weed out co-factors.
+        options = []
+        for compound in substrates:
+            # Generate the atom counts from the formulae.
+            count = atom_counts.setdefault(compound,
+                                           count_atoms(compound.formula))
+            target_count = atom_counts.setdefault(target,
+                                                  count_atoms(target.formula))
+            scl = compute_chemical_linkage_strength(
+                count, target_count)
+            options.append((compound, scl))
+        # The maximum SCL is our candidate precursor.
+        options.sort(key=itemgetter(1))
```

Comment:
Suggestion: `max(options, key=itemgetter(1))` would perhaps describe the intention here even more clearly.

Reply:
I decided to keep it in order to have the logging output sorted.